opnode/node/driver.go

```diff
@@ -1,40 +1,79 @@
 package node
 
 import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/ethereum/go-ethereum/trie"
 	"github.com/holiman/uint256"
 )
 
+const (
+	DepositTxTypePrefix = 0x7E
+)
+
+var (
+	ErrTxTypePrefix     = errors.New("wrong transaction type prefix")
+	DepositContractAddr = common.HexToAddress("0xdeaddeaddeaddeaddeaddeaddeaddeaddead0001")
+	BatchSubmissionAddr = common.HexToAddress("0x0000000000000000000000006f7074696d69736d") // TODO pick address
+	SequencerAddr       = common.HexToAddress("0x4242424242424242424242424242424242424242") // TODO pick address
+)
+
 type InputItem struct {
 	Block    *types.Block
 	Receipts []*types.Receipt
 }
 
-func (item *InputItem) Epoch() uint64 {
-	return item.Block.NumberU64()
-}
-
 // sanity check that the receipts are consistent with the block data
 func (item *InputItem) CheckReceipts() bool {
 	hasher := trie.NewStackTrie(nil)
 	computed := types.DeriveSha(types.Receipts(item.Receipts), hasher)
 	return item.Block.ReceiptHash() == computed
 }
 
+// NOTE: It's impossible to define DecodeRLP on a pointer of pointer, so we need the struct.
+//       Defining DecodeRLP on DepositTransaction is too much of a hassle due to manual error handling.
+type AddressOrNil struct {
+	Ptr *common.Address
+}
+
+func (addr *AddressOrNil) DecodeRLP(s *rlp.Stream) error {
+	var bytes []byte
+	var err error
+	if bytes, err = s.Bytes(); err != nil {
+		return err
+	} else if len(bytes) == 0 {
+		addr.Ptr = nil
+		return nil
+	} else if len(bytes) != common.AddressLength {
+		// we can't use `s.Decode(addr.Ptr)` because we already read the bytes
+		return fmt.Errorf("rlp: wrong address size: %d", len(bytes))
+	} else {
+		addr.Ptr = new(common.Address)
+		copy(addr.Ptr[:], bytes)
+		return nil
+	}
+}
+
 type DepositTransaction struct {
-	Sender common.Address
-	// TODO some flag/bool for contract-deploy transaction. Or it can be *To with nil
-	Flags uint8
-	To    common.Address
-	Value uint256.Int
-	Data  []byte
+	From     common.Address
+	To       AddressOrNil // nil for the empty/nil address (∅)
```

Comment:
This can just be a one-liner, see upstream go-ethereum implementation of the regular tx type with optional `To` field:
https://github.com/ethereum/go-ethereum/blob/4e599ee469223ccc452f40f36c85bac3bea9c16b/core/types/dynamic_fee_tx.go#L31